### PR TITLE
Beginnings of an Alertmanager mixin.

### PR DESCRIPTION
**Pull request by tomwilkie**
Alerts are from https://github.com/coreos/prometheus-operator/blob/master/contrib/kube-prometheus/jsonnet/kube-prometheus/alerts/ and https://github.com/prometheus/prometheus/pull/4474.

Signed-off-by: Tom Wilkie <tom.wilkie@gmail.com>

### CODE REVIEW

**Comment by mxinden** [comment]
@tomwilkie first of all thanks for pushing this forward.

Might it make sense to add a `config.libsonnet` like in https://github.com/prometheus/prometheus/pull/4474?

**Comment by tomwilkie** [comment]
@beorn7 can you add this to your TODO list please?

**Comment by beorn7** [comment]
I have done that long ago. Just that there are a lot of items on that list…

**Comment by simonpasquier** [comment]
@beorn7 @tomwilkie I'm ok to tackle this one if you don't mind...

**Comment by beorn7** [comment]
Yes, please go ahead. I have this on my long todo list, but way too far down.
Any help appreciated.

**Comment by beorn7** [comment]
@simonpasquier I have some practical reasons to work on this. Will push a few commits today and tomorrow.

**Comment by roidelapluie** [comment]
> @simonpasquier I have some practical reasons to work on this. Will push a few commits today and tomorrow.

Oh, I was going to say the exact same thing. Happy to try & review.

**Comment by beorn7** [comment]
OK, got all my piled-up ideas implemented. I think this is meaty enough now to get merged into the main branch.

Reviewers, please have another look.

@duologic this might be of special interest to you. We could try this out from this branch.

File: doc/alertmanager-mixin/alerts.libsonnet [hunk 1]
```diff
@@ -0,0 +1,135 @@
...
+                avg_over_time(up{%(alertmanagerSelector)s}[5m])
+              ) < 0.5
+            ||| % $._config,
+            'for': '5m',
+            labels: {
+              severity: 'critical',
+            },
+            annotations: {
+              summary: 'All Alertmanager instances within the same cluster are down.',
+              description: 'Each Alertmanager instances within the %(alertmanagerClusterName)s cluster has been up for less than {{ $value | humanizePercentage }} of the last 5m.' % $._config,
```
Comment (roidelapluie) [inline doc/alertmanager-mixin/alerts.libsonnet:112]: Should'nt we alert earlier? if they are all down it's too late to get the notification :) as a reminder it is regularly said that one alertmanager cluster is enough company-wide

Comment (beorn7) [reply doc/alertmanager-mixin/alerts.libsonnet:112]: Right. Let me rephrase this to alert if half of the instances are down.

In any case, a proper alerting setup should also have an end-to-end test with an always firing alert, which, if it doesn't call in anymore, will trigger a secondary alert via a service like Dead Man's Snitch.

Comment (beorn7) [reply doc/alertmanager-mixin/alerts.libsonnet]: OK, done. The alerts fire now if half or more of the cluster is affected, which means 2 or more with a cluster of 4 instances, which means that you need two failed instances (rather than only one) before getting a page.

**Comment by beorn7** [comment]
All looks good so far.

@simonpasquier I think this should be considered for merging. Could you do a final review?

**Comment by beorn7** [comment]
@simonpasquier are you available to review this? Or alternatively, would you be fine delegating review to someone else?

**Comment by beorn7** [comment]
@simonpasquier or perhaps @brancz @tomwilkie @gouthamve @roidelapluie ? I guess this is of some interest to all of you.

I'd really like to get any feedback on this and then merge this finally.

File: doc/alertmanager-mixin/Makefile [hunk 2]
```diff
@@ -0,0 +1,22 @@
...
+$(ALERTMANAGER_ALERTS): mixin.libsonnet alerts.libsonnet
+	jsonnet -S alerts.jsonnet > $@
+
+lint: $(ALERTMANAGER_ALERTS)
+	find . -name 'vendor' -prune -o -name '*.libsonnet' -print -o -name '*.jsonnet' -print | \
+		while read f; do \
+			$(JSONNET_FMT) "$$f" | diff -u "$$f" -; \
+		done
+
+	promtool check rules $(ALERTMANAGER_ALERTS)
```
Comment (tomwilkie) [inline doc/alertmanager-mixin/Makefile:19]: Do you mind if we use mixtool here?  Happy to make the change myself.

**Review by tomwilkie — COMMENTED** [review]
LGTM from me, with a minor nit.

File: doc/alertmanager-mixin/README.md [hunk 3]
```diff
@@ -0,0 +1,41 @@
...
+
+To use them, you need to have `jsonnet` (v0.13+) and `jb` installed. If you
+have a working Go development environment, it's easiest to run the following:
+
+```bash
+$ go get github.com/google/go-jsonnet/cmd/jsonnet
+$ go get github.com/jsonnet-bundler/jsonnet-bundler/cmd/jb
+```
+
+_Note: The make targets `lint` and `fmt` need the `jsonnetfmt` binary, which is
```
Comment (roidelapluie) [inline doc/alertmanager-mixin/README.md:22]: This is no lomger true with 1.16

File: doc/alertmanager-mixin/README.md [hunk 4]
```diff
@@ -0,0 +1,41 @@
...
+
+The alerts are designed to monitor a cluster of Alertmanager instances. To make
+them work as expected, the Prometheus server the alerts are evaluated on has to
+scrape all Alertmanager instances of the cluster, even if those instances are
+distributed over different locations. All Alertmanager instances in the same
+Alertmanager cluster must have the same `job` label. In turn, if monitoring
+multiple different Alertmanager clusters, instances from different clusters
+must have a different `job` label.
+
+To use them, you need to have `jsonnet` (v0.13+) and `jb` installed. If you
```
Comment (roidelapluie) [inline doc/alertmanager-mixin/README.md:14]: 0.16 so we have jsonnetfmt

File: doc/alertmanager-mixin/README.md [hunk 5]
```diff
@@ -0,0 +1,41 @@
...
+have a working Go development environment, it's easiest to run the following:
+
+```bash
+$ go get github.com/google/go-jsonnet/cmd/jsonnet
```
Comment (roidelapluie) [inline doc/alertmanager-mixin/README.md:18]:
```suggestion
$ go get github.com/google/go-jsonnet/cmd/jsonnetfmt
```

**Comment by roidelapluie** [comment]
Can you add the CI like we have in Prometheus?

**Comment by beorn7** [comment]
Great feedback. Thank you very much. I'll incorporate it tomorrow (hopefully…).

**Comment by brancz** [comment]
lgtm :+1:

File: doc/alertmanager-mixin/Makefile (hunk 2 above)
Comment (beorn7) [reply doc/alertmanager-mixin/Makefile:19]: Done.

**Comment by beorn7** [comment]
Sorry all for the delay, had busy weeks recently. Finally addressed all the comments.

Getting `promtool` to build within CI was PITA because of that dreaded clash between semantic versioning we use for Prometheus releases and how Go modules insists on using it on the code level.

**Comment by roidelapluie** [comment]
Mmh Makefile.common knows how to download a compiled version of promtool.

**Comment by beorn7** [comment]
> Makefile.common knows how to download a compiled version of promtool.

I couldn't find that. Could you explain in more detail please?

**Comment by roidelapluie** [comment]
We have a $(PROMU) target that downloads promu, you do not need to go get it.

**Comment by beorn7** [comment]
But we don't need PROMU, we need promtool.

And getting promu to then compile promtool seems even worse than what I'm doing now.

At the very least, I want to keep the Makefile simple because it is not the Makefile of Alertmanager, it's just a simple Makefile for users of the mixin. They really shouldn't get in touch with Promu.

**Comment by beorn7** [comment]
If you know a way to put this all into the CircleCI config without bloating the makefile, please let me know.

**Comment by roidelapluie** [comment]
Sorry, I made the confusion between promu and promtool.

It looks like you could look at
https://github.com/monitoring-mixins/mixtool too

**Comment by beorn7** [comment]
> It looks like you could look at https://github.com/monitoring-mixins/mixtool too

That's already in here. Am I missing anything?

**Comment by beorn7** [comment]
@roidelapluie do you have any merge-blocking concerns left? I'd like to get this in if there is nothing substantially wrong with it. We can still improve CI details later.

**Comment by beorn7** [comment]
Now I see that `mixtool` seems to include the `promtool check rules` capability. That's not documented, and I had a short look at the code and couldn't see it, but it's in there. So I can rip out `promtool` altogether.

**Review by roidelapluie — APPROVED** [review]
Thanks, LGTM!!

**Comment by beorn7** [comment]
Thanks, everyone. Merging now (and will create a new PR with a new feature momentarily, I just didn't want to load up this here even more).

Because we have accumulated 20 commits in this PR, I will make an exception from my usual practice and squash them.
